src/lib/tzcode/timezone.c

```diff
@@ -96,27 +115,117 @@ char_span_alpha(const char *src, size_t len)
 	return n;
 }
 
-ssize_t
-timezone_lookup(const char *str, size_t len, const struct date_time_zone **zone)
+static inline ssize_t
+timezone_raw_lookup(const char *str, size_t len,
+		    struct date_time_zone **zone)
 {
 	len = char_span_alpha(str, len);
 	if (len == 0)
 		return 0;
 
-	struct date_time_zone *sorted = sort_abbrevs_singleton();
 	struct date_time_zone wrap = {.name = str};
 	struct date_time_zone *found = (struct date_time_zone *)
-		bsearch(&wrap, sorted, lengthof(zone_abbrevs),
-			sizeof(struct date_time_zone), compare_abbrevs);
+		bsearch(&wrap, zones_sorted, lengthof(zones_sorted),
+			sizeof(zones_sorted[0]), compare_zones);
 
 	if (found != NULL) {
 		/* lua assumes that single bit is set, not both */
-		assert((found->flags & (TZ_NYI | TZ_AMBIGUOUS)) !=
-		       (TZ_NYI | TZ_AMBIGUOUS));
-		if (found->flags & (TZ_NYI | TZ_AMBIGUOUS))
-			return -found->flags;
+		assert((found->flags & TZ_ERROR_MASK) != TZ_ERROR_MASK);
+		if (found->flags & TZ_ERROR_MASK)
+			return -(found->flags & TZ_ERROR_MASK);
 		*zone = found;
 		return len;
 	}
 	return -1;
 }
+
+ssize_t
+timezone_tm_lookup(const char *str, size_t len,
+		   struct date_time_zone **zone,
+		   struct tnt_tm *tm)
+{
+	ssize_t rc = timezone_raw_lookup(str, len, zone);
+	if (rc <= 0)
+		return rc;
+
+	const struct date_time_zone *found = *zone;
+	if ((found->flags & TZ_OLSON) == 0) {
+		tm->tm_gmtoff = found->offset * 60;
+		tm->tm_tzindex = found->id;
+		tm->tm_isdst = !!(found->flags & TZ_DST);
+		return rc;
+	}
+	timezone_t tz = tzalloc(str); // FIXME - cache loaded
+	if (tz == NULL)
+		return 0;
+	struct datetime date = {.epoch = 0};
+	if (tm_to_datetime(tm, &date) == false)
+		goto exit_0;
+	time_t epoch = (int64_t)date.epoch;
+	struct tnt_tm * result = tnt_localtime_rz(tz, &epoch, tm);
+	if (result == NULL)
+		goto exit_0;
+	tzfree(tz);
+	return rc;
+exit_0:
+	if (tz != NULL)
+		tzfree(tz);
+	return 0;
+}
+
+ssize_t
+timezone_epoch_lookup(const char *str, size_t len, time_t base,
+		      struct date_time_zone **zone)
+{
+	ssize_t rc = timezone_raw_lookup(str, len, zone);
+	if (rc <= 0)
+		return rc;
+
+	struct date_time_zone *found = *zone;
+	if (found->flags & (TZ_RFC | TZ_UTC | TZ_MILITARY)) {
+		assert((TZ_ERROR_MASK & found->flags) == 0);
+		return rc;
+	}
+	timezone_t tz = NULL;
+	if (found->flags & TZ_OLSON) {
+		tz = tzalloc(str); // FIXME - cache loaded
+		if (tz == NULL)
+			return 0;
+		struct tnt_tm tm = {.tm_epoch = 0};
+		struct tnt_tm * result = tnt_localtime_rz(tz, &base, &tm);
+		if (result == NULL)
+			goto exit_0;
+		// FIXME - found a better way to return tzoffset
+		found->offset = result->tm_gmtoff / 60;
+		found->flags &= ~TZ_DST;
+		found->flags |= TZ_DST * result->tm_isdst;
+		tzfree(tz);
+	}
+	return rc;
+exit_0:
+	if (tz != NULL)
+		tzfree(tz);
+	return 0;
+}
+
+bool
+timezone_tzindex_lookup(int16_t tzindex, struct tnt_tm *tm)
+{
+	assert(tm != NULL);
+	if (tzindex == 0)
+		return true;
```

Comment:
Because we could return nothing useful if there is 0 in `tzindex` we are looking for. (i.e. there is 0-sideeffect but it's still not and error we should care about thus `return true`)

Reply:
I mean, is it even possible that tzindex is 0 in `timezone_tzindex_lookup`?
If it is, you should return false IMO.

Imagine you're looking at this from outside code, and think that `timezone_tzindex_lookup` worked fine, because it returned true. Now you try to use struct tnt_tm, but it's uninitialized, because tzindex was 0.

In other words in my opinion tzindex 0 should be "not found" like any other invalid tzindex

Reply:
Ok, agreed. Changed to return 'false' if we did nothing here.